File: src/main/java/model/Point.java
package model;

import lombok.AllArgsConstructor;
import lombok.Data;

@Data
@AllArgsConstructor
public class Point {
    private final double x;
    private final double y;

    /**
     * Reflect point across {@link Line} using formula:
     * https://math.stackexchange.com/a/1743581
     *
     * Handle horizontal and vertical lines seperately because that formula didn't cover those cases.
     *
     * @param mirrorLine to reflect point across.
     * @return reflected point. Returns itself if point is on the line.
     */
    public Point reflect(Line mirrorLine) {
        if (this.satisfiesLine(mirrorLine)) {
            return this;
        }

        if (mirrorLine.isHorizontal()) {
            final double yIntercept = -1 * (mirrorLine.getC() / mirrorLine.getB());
            final double yDelta = Math.abs(yIntercept - y);
            return y < yIntercept ? new Point(x, yIntercept + yDelta) : new Point(x, yIntercept - yDelta);
        }

        if (mirrorLine.isVertical()) {
            final double xIntercept = -1 * (mirrorLine.getC() / mirrorLine.getA());
            final double xDelta = Math.abs(xIntercept - x);
            return x < xIntercept ? new Point(xIntercept + xDelta, y) : new Point(xIntercept - xDelta, y);
        }

        final double aSquared = mirrorLine.getA() * mirrorLine.getA();
        final double bSquared = mirrorLine.getB() * mirrorLine.getB();
        final double denominator = aSquared + bSquared;

        double reflectedX = x * (aSquared - bSquared);
        reflectedX -= 2 * mirrorLine.getB() * ((mirrorLine.getA() * y) + mirrorLine.getC());
        reflectedX /= denominator;

        double reflectedY = y * (bSquared - aSquared);
        reflectedY -= 2 * mirrorLine.getA() * ((mirrorLine.getB() * x) + mirrorLine.getC());
        reflectedY /= denominator;

        return new Point(reflectedX, reflectedY);
    }

    /**
     * Check if point sits on given {@link Line}.
     * Checks if point satisfies equation ax + by + c = 0.
     *
     * @param line to test point against.
     * @return true if point is on line.
     */
    public boolean satisfiesLine(Line line) {
        final double sum = (line.getA() * x) + (line.getB() * y) + line.getC();
        return sum == 0;
    }

    /**
     * Get midpoint between two points.
     */
    public static Point getMidpoint(Point p1, Point p2) {
        final double midpointX = (p1.getX() + p2.getX()) / 2;
        final double midpointY = (p1.getY() + p2.getY()) / 2;
        return new Point(midpointX, midpointY);
    }

}

File: src/main/java/model/Point.kt
package model

data class Point(val x: Double, val y: Double) {

    /**
     * Reflect point across [Line] using formula: https://math.stackexchange.com/a/1743581
     */
    fun reflect(lineOfReflection: Line): Point {
        if (satisfiesLine(lineOfReflection)) return this

        val aSquared = lineOfReflection.a * lineOfReflection.a
        val bSquared = lineOfReflection.b * lineOfReflection.b
        val denominator = aSquared + bSquared

        var reflectedX = x * (bSquared - aSquared)
        reflectedX -= 2 * lineOfReflection.a * (lineOfReflection.b * y + lineOfReflection.c)
        reflectedX /= denominator

        var reflectedY = y * (aSquared - bSquared)
        reflectedY -= 2 * lineOfReflection.b * (lineOfReflection.a * x + lineOfReflection.c)
        reflectedY /= denominator

        return Point(reflectedX, reflectedY)
    }

    /**
     * Check if point sits on given [Line].
     */
    fun satisfiesLine(line: Line): Boolean {
        val sum = line.a * x + line.b * y + line.c
        return sum == 0.0
    }

    companion object {
        fun getMidpoint(p1: Point, p2: Point): Point {
            val midpointX = (p1.x + p2.x) / 2
            val midpointY = (p1.y + p2.y) / 2
            return Point(midpointX, midpointY)
        }
    }
}